src/background/javascript-instrument.ts

```diff
@@ -5,63 +5,113 @@ import { boolToInt, escapeString, escapeUrl } from "../lib/string-utils";
-      update.symbol = escapeString(data.symbol);
-      update.operation = escapeString(data.operation);
-      update.value = escapeString(data.value);
-      update.time_stamp = data.timeStamp;
-      update.incognito = boolToInt(sender.tab.incognito);
-
-      // document_url is the current frame's document href
-      // top_level_url is the top-level frame's document href
-      update.document_url = escapeUrl(sender.url);
-      update.top_level_url = escapeUrl(sender.tab.url);
-
-      if (data.operation === "call" && data.args.length > 0) {
-        update.arguments = escapeString(JSON.stringify(data.args));
+  /**
+   * Start listening for messages from page/content/background scripts injected to instrument JavaScript APIs
+   */
+  public listen() {
+    this.onMessageListener = (message, sender) => {
+      // console.debug("javascript-instrumentation background listener", {message, sender}, this.configured);
+      if (
+        message.namespace &&
+        message.namespace === "javascript-instrumentation"
+      ) {
+        this.handleJsInstrumentationMessage(message, sender);
       }
-
-      this.dataReceiver.saveRecord("javascript", update);
     };
+    browser.runtime.onMessage.addListener(this.onMessageListener);
+  }
 
-    // Listen for messages from content script injected to instrument JavaScript API
-    this.onMessageListener = (msg, sender) => {
-      // console.debug("javascript-instrumentation background listener - msg, sender, sendReply", msg, sender, sendReply);
-      if (msg.namespace && msg.namespace === "javascript-instrumentation") {
-        switch (msg.type) {
-          case "logCall":
-          case "logValue":
-            processCallsAndValues(msg.data, sender);
-            break;
+  /**
+   * Either sends the log data to the dataReceiver or store it in memory
+   * as a pending record if the JS instrumentation is not yet configured
+   * @param message
+   * @param sender
+   */
+  public handleJsInstrumentationMessage(message, sender: MessageSender) {
+    switch (message.type) {
+      case "logCall":
+      case "logValue":
+        const update = JavascriptInstrument.processCallsAndValues(
+          message.data,
+          sender,
+        );
+        if (this.configured) {
+          update.crawl_id = this.crawlID;
+          this.dataReceiver.saveRecord("javascript", update);
+        } else {
+          this.pendingRecords.push(update);
         }
-      }
-    };
-    browser.runtime.onMessage.addListener(this.onMessageListener);
+        break;
+    }
+  }
+
+  /**
+   * Starts listening if haven't done so already, sets the crawl ID,
+   * marks the JS instrumentation as configured and sends any pending
+   * records that have been received up until this point.
+   * @param crawlID
+   */
+  public run(crawlID) {
+    if (!this.onMessageListener) {
+      this.listen();
+    }
+    this.crawlID = crawlID;
+    this.configured = true;
+    this.pendingRecords.map(update => {
```

Comment:
Is `this.pendingRecords` ever initialized?

Reply:
Yes, initialized [here](https://github.com/mozilla/openwpm-webext-instrumentation/pull/44/files/c8e9da25a0217b562b31bdbfd054114b778e9cdb#diff-41d9a188b7078bfcc73766acbfe45e63R48) and used [here](https://github.com/mozilla/openwpm-webext-instrumentation/pull/44/files/c8e9da25a0217b562b31bdbfd054114b778e9cdb#diff-41d9a188b7078bfcc73766acbfe45e63R89)